KEYCLOAK-17887 fix PATCH endpoint for realm localization

danielFesenmeyer opened the pull request:
UnsupportedOperationException was thrown because RealmAdapter tried to change unmodifiable map. Fixed this by removing unnecessary line of code.

To prevent regression, I added tests for the PATCH endpoint.

danielFesenmeyer committed:
KEYCLOAK-17887 fix endpoint for creating or updating realm localization texts for a given locale (UnsupportedOperation was thrown because RealmAdapter tried to change unmodifiable map):
- fix RealmAdapter to create a new map instead of trying to change unmodifiable map
- only provide POST endpoints for creating or updating the texts (to have the endpoints consistent with other Admin API endpoints)
- add tests

stianst reviewed (changes requested):
I appreciate that PATCH should be used in this case, but we don't use it anywhere else in the Admin APIs, so this would be rather inconsistent with other things.

What we do instead is somewhat implement patch as part of the post, which I appreciate is quite broken, but think that's what we should do here as well to make it consistent with other endpoints.

danielFesenmeyer commented:
> I appreciate that PATCH should be used in this case, but we don't use it anywhere else in the Admin APIs, so this would be rather inconsistent with other things.
> 
> What we do instead is somewhat implement patch as part of the post, which I appreciate is quite broken, but think that's what we should do here as well to make it consistent with other endpoints.

Ok, I can change the corresponding endpoints to POST. Is it ok that this breaks the API, or should the PATCH endpoints be kept for backwards compatibility?

danielFesenmeyer commented:
> > I appreciate that PATCH should be used in this case, but we don't use it anywhere else in the Admin APIs, so this would be rather inconsistent with other things.
> > What we do instead is somewhat implement patch as part of the post, which I appreciate is quite broken, but think that's what we should do here as well to make it consistent with other endpoints.
> 
> Ok, I can change the corresponding endpoints to POST. Is it ok that this breaks the API, or should the PATCH endpoints be kept for backwards compatibility?

@stianst WDYT?

danielFesenmeyer commented:
I have changed the implementation to use POST instead of PATCH. I also renamed RealmModel#patchRealmLocalizationTexts to #createOrUpdateRealmLocalizationTexts, in order to avoid confusion.

I didn't add any code for backwards compatibility, because the endpoint was undocumented anyway (not in org.keycloak.admin.client.resource.RealmLocalizationResource). Please let be me know if you think that backwards compatibility is needed.

stianst commented:
Rebased here: https://github.com/keycloak/keycloak/pull/8509

stianst commented:
Merged in https://github.com/keycloak/keycloak/pull/8509
